Custom themes

guy-black opened the pull request:
- learned more about git so that future work I do on the source won't keep showing up in this pull request this time
- added a -dct <path/to/file> flag to load a custom theme file
- added two example theme files one heavily commented
- removed the read and show instances on EdDesign and instead used PrettyEdDesign datatype
-  modified Argument.hs to read the theme file as PrettyEdDesign and then convert to Ed Design before updating the config

guy-black committed:
-derived read and show instances to EdDesign
-worked out how to read a plaintext theme file with comments from disk and read as an EdDesign
-added external plaintext theme files

guy-black committed:
added a -dct <path/to/theme> option
tested that it successfully loads external theme file
have not done anything to handle exceptions if the file isn't there or can't be read into EdDesign yet
livin dangerous for a bit here

guy-black committed:
Merge branch 'LadyBoonami:master' into master

guy-black committed:
theme not found error working fine.

theme not reading not quite

guy-black committed:
Merge branch 'master' of github.com:guy-black/wsedit

guy-black committed:
my bad it works now

guy-black committed:
oops

guy-black committed:
tfw it was all working and then after I finished writing helpful comments the default theme decides to not work and I just

guy-black committed:
now it all works right

guy-black committed:
...

guy-black committed:
- fixed up my git mess and learned the value of branches
- removed the ```Read``` and ```Show``` instances from ```EdDesign```
- modified the themes files to be read in as ```PrettyEdDesign```
- modified the ```applyArg``` function to call ```unPrettyEdDesign``` on the
	```PrettyEdDesign``` pulled from theme file
- somewhere down the line I want to see if i can modify the ```PrettyEdDesign```
 	data type and ```Read``` and ```Write``` Instances to make editing
 	the theme file a bit more pleasant

LadyBoonami commented:
Wow! At first glance, this is pretty much exactly what I had in mind. Thank you very much for your contribution =)

Sadly I'm pretty busy this and probably next week, so I can't promise when I'll have the time to take a good look at this PR...

guy-black commented:
No worries!  Please feel free to leave whatever feedback or criticism you want, I've just been teaching myself and hope to eventually be good enough to get a job :)

LadyBoonami commented:
Alright, this looks pretty solid. I'm especially happy with how you kept the indentation consistent :heart:

But it's not quite ready to be merged yet. There are a few (small) general things that I'm not quite happy with, some failures need to be handled more gracefully, and then there are a few other (very minor) issues:

In General:
- All the other options use one letter per "thing" they select, so e.g. one letter for the category, one letter for the general option, and possibly one letter to differentiate between variants. The `-dct` switch is not consistent with that naming scheme, I'd prefer something like `-dt` or `-dc`.
- Also, having a `-dT` option to reset the theme to the default would be nice.
- Currently, the editor crashes with an error message if the theme file cannot be opened or parsed. The parse error message could provide a bit more information (e.g. what should I do? Which file could not be parsed?). I'm aware that `read` does not produce useful error messages, so it's okay to not have the exact error location in the message for now. It's probably also a good idea to tell people to check the changelog at `WSEdit.Data.upstream ++ "/blob/master/CHANGES.md"` if things broke recently without them changing anything.
- More importantly, `-mf` does nothing for both error conditions. This means you need to either move files around or use another editor to fix any errors, which should never happen with any dignified editor. Instead, any theme file errors should be ignored if `-mf` is given, and the error message should hint at using `-mf` to correct problems.
- An extension and a syntax definition for theme files would be great to help with errors. My pick would be `.wstheme`, but feel free to choose something appropriate.
- You mention "file route" a few times in the code. Did you mean file *path*?

Implementation Details:
- The `LANGUAGE BinaryLiterals` directives in `src/WSEdit/Arguments.hs` and `src/WSEdit/Data.hs` don't seem to be doing anything, or am I missing something?
- `src/WSEdit/Arguments.hs:315`: `head` is an incomplete function that will `error` on empty lists (i.e. empty lines in the input). Using something like `take 1` prevents ugly things from happening.
- The box in `theme/default:19` is a bit misaligned.
- The option in `src/WSEdit/Help.hs:344` is incorrect.
- The build is throwing warnings for unused imports, which breaks the debug build (`stack build --flag wsedit:dev`, which builds a separate executable named `wsed-dev`) due to `-Werror`.

Finally: please bump the last version digit in `wsedit.cabal` and `src/WSEdit/Data.hs` and create a `CHANGES.md` entry in the "Unreleased" category.

Apart from that, it's looking great, please do keep up the good work :+1:

guy-black commented:
Thank you so much for your feedback!
To answer your questions real quick, yeah I mean file path.  I'm still a little rough on the proper terminology here and there, thanks for pointing that out.
And the BinaryLiterals I totally forgot to take back out, oops :sweat_smile: the original idea there was to be able to write the Vty Style as a binary literal instead of having to convert it to hex or dec later, but I couldn't get it to parse.  I scrapped the idea since I'm already thinking I'd like to either learn the dark arts of parsec, or fiddle with the ```read```  instance of PrettyEdDesign to make the config file a bit more neat, and give better error messages if it can't be parsed.

Thank you again for the feedback!
I really appreciate it and am looking forward to getting on to making those changes soon :smiley:

LadyBoonami commented:
The large benefit of `Read` is that code generation is completely automated, including future updates to `EdDesign`. But for that, you are bound to a very rigid format (you can't even change the order of fields) and get no feedback except "Prelude.read: no parse." Additionally, all existing theme files will break whenever `EdDesign` is updated. This tradeoff is fine to get initial theming support into WSEdit without too much effort, but it's obviously not optimal.

The "ultimate form" of theming I can think of would probably be something like an option `-dt <name> <attr>` where `<name>` is the name of a field in `EdDesign` and `<attr>` is the value (to be processed by `read`) to overwrite it with (also `-dT <name>` to reset). That would reuse the existing config file format, get us theme cascading for free (e.g. red frame when root), be much more robust to implementation changes, and make errors much more fine-grained. But that's obviously a lot more work, and I don't know of any way to get around writing a parse rule for every current and future field of `EdDesign`.

So for now, a `Read` based implementation is probably fine.

Oh, and by the way: I just figured out what the `attrURL` field does: https://hackage.haskell.org/package/vty-5.33/docs/Graphics-Vty-Attributes.html#v:withURL

guy-black committed:
I regret to inform you all that I am once again writing code regularly

guy-black committed:
anotherone

guy-black committed:
got basically all the easy things done

guy-black committed:
added wstheme.wsconf

guy-black committed:
fix

guy-black committed:
changed flag to dt, added dT
changed displayCustTheme/DisplayCustTheme to displayThemeOn/DisplayThemeOn
added displayThemeOff/DisplayThemeOff

guy-black committed:
added better complaints if themefile can't be loaded
if theme file can't be parsed and -mf flag passed in, ignore theme file

guy-black committed:
fixed -mf so it will ignore if a theme file can't be found either

guy-black committed:
added updated `CHANGES.md`, bumped version number, updated URL explanation in example theme file

guy-black committed:
preparing to merge back into custom-themes

guy-black commented:
Sorry I kinda disappeared for a while there!  I finally got through to finishing up those changes.
tl;dr
changed ```-dct``` to ```-dt```
added ```-dT``` to ignore external theme
wrote slightly more detailed complaints for theme files that can't be found or read
```-mf``` will ignore any theme files that can't be found or read
also added a ```wstheme.wsconf``` to ```lang/```

guy-black committed:
Merge branch 'master' of https://github.com/LadyBoonami/wsedit into LadyBoonami-master

guy-black committed:
bumped version info

guy-black committed:
i need to git gud

LadyBoonami commented:
Alright, sorry for keeping you waiting for so long. This looks pretty good, so I'll merge it with some very minor changes, and create a new release in a few days if nothing blows up until then. Thanks a lot for contributing =)
